From 27020d86583bb391221c46c9e7686126b81e4456 Mon Sep 17 00:00:00 2001
From: Rashed Abdel-Tawab <rashed@linux.com>
Date: Thu, 19 Jul 2018 14:17:59 -0700
Subject: [PATCH 5/9] interfaces: Add vendor.lineage.stache@1.0::ISecureStorage

Whee time for some secure storage. This will leverage EXT4's
file encryption to create encrypted containers in /data/stache.
This is just the basebones HIDL HAL definition for now, the
implementation will come in a later commit.

Change-Id: I09911ea0dcb42ce0771d734673f117f2cb0c331b
---
 stache/1.0/Android.bp         |  89 +++++++++++++++++++++++++
 stache/1.0/Android.mk         | 118 ++++++++++++++++++++++++++++++++++
 stache/1.0/ISecureStorage.hal |  24 +++++++
 stache/1.0/types.hal          |  26 ++++++++
 stache/Android.bp             |   4 ++
 5 files changed, 261 insertions(+)
 create mode 100644 stache/1.0/Android.bp
 create mode 100644 stache/1.0/Android.mk
 create mode 100644 stache/1.0/ISecureStorage.hal
 create mode 100644 stache/1.0/types.hal
 create mode 100644 stache/Android.bp

diff --git a/stache/1.0/Android.bp b/stache/1.0/Android.bp
new file mode 100644
index 0000000..f1c9461
--- /dev/null
+++ b/stache/1.0/Android.bp
@@ -0,0 +1,89 @@
+// This file is autogenerated by hidl-gen. Do not edit manually.
+
+filegroup {
+    name: "vendor.lineage.stache@1.0_hal",
+    srcs: [
+        "types.hal",
+        "ISecureStorage.hal",
+    ],
+}
+
+genrule {
+    name: "vendor.lineage.stache@1.0_genc++",
+    tools: ["hidl-gen"],
+    cmd: "$(location hidl-gen) -o $(genDir) -Lc++-sources -randroid.hidl:system/libhidl/transport -rvendor.lineage:hardware/lineage/interfaces vendor.lineage.stache@1.0",
+    srcs: [
+        ":vendor.lineage.stache@1.0_hal",
+    ],
+    out: [
+        "vendor/lineage/stache/1.0/types.cpp",
+        "vendor/lineage/stache/1.0/SecureStorageAll.cpp",
+    ],
+}
+
+genrule {
+    name: "vendor.lineage.stache@1.0_genc++_headers",
+    tools: ["hidl-gen"],
+    cmd: "$(location hidl-gen) -o $(genDir) -Lc++-headers -randroid.hidl:system/libhidl/transport -rvendor.lineage:hardware/lineage/interfaces vendor.lineage.stache@1.0",
+    srcs: [
+        ":vendor.lineage.stache@1.0_hal",
+    ],
+    out: [
+        "vendor/lineage/stache/1.0/types.h",
+        "vendor/lineage/stache/1.0/hwtypes.h",
+        "vendor/lineage/stache/1.0/ISecureStorage.h",
+        "vendor/lineage/stache/1.0/IHwSecureStorage.h",
+        "vendor/lineage/stache/1.0/BnHwSecureStorage.h",
+        "vendor/lineage/stache/1.0/BpHwSecureStorage.h",
+        "vendor/lineage/stache/1.0/BsSecureStorage.h",
+    ],
+}
+
+cc_library {
+    name: "vendor.lineage.stache@1.0",
+    defaults: ["hidl-module-defaults"],
+    generated_sources: ["vendor.lineage.stache@1.0_genc++"],
+    generated_headers: ["vendor.lineage.stache@1.0_genc++_headers"],
+    export_generated_headers: ["vendor.lineage.stache@1.0_genc++_headers"],
+    vendor_available: true,
+    vndk: {
+        enabled: true,
+    },
+    shared_libs: [
+        "libhidlbase",
+        "libhidltransport",
+        "libhwbinder",
+        "liblog",
+        "libutils",
+        "libcutils",
+    ],
+    export_shared_lib_headers: [
+        "libhidlbase",
+        "libhidltransport",
+        "libhwbinder",
+        "libutils",
+    ],
+}
+
+cc_library {
+    name: "vendor.lineage.stache@1.0_vendor",
+    defaults: ["hidl-module-defaults"],
+    generated_sources: ["vendor.lineage.stache@1.0_genc++"],
+    generated_headers: ["vendor.lineage.stache@1.0_genc++_headers"],
+    export_generated_headers: ["vendor.lineage.stache@1.0_genc++_headers"],
+    vendor: true,
+    shared_libs: [
+        "libhidlbase",
+        "libhidltransport",
+        "libhwbinder",
+        "liblog",
+        "libutils",
+        "libcutils",
+    ],
+    export_shared_lib_headers: [
+        "libhidlbase",
+        "libhidltransport",
+        "libhwbinder",
+        "libutils",
+    ],
+}
diff --git a/stache/1.0/Android.mk b/stache/1.0/Android.mk
new file mode 100644
index 0000000..2a501c1
--- /dev/null
+++ b/stache/1.0/Android.mk
@@ -0,0 +1,118 @@
+# This file is autogenerated by hidl-gen. Do not edit manually.
+
+LOCAL_PATH := $(call my-dir)
+
+################################################################################
+
+include $(CLEAR_VARS)
+LOCAL_MODULE := vendor.lineage.stache-V1.0-java
+LOCAL_MODULE_CLASS := JAVA_LIBRARIES
+
+intermediates := $(call local-generated-sources-dir, COMMON)
+
+HIDL := $(HOST_OUT_EXECUTABLES)/hidl-gen$(HOST_EXECUTABLE_SUFFIX)
+
+LOCAL_JAVA_LIBRARIES := \
+    android.hidl.base-V1.0-java \
+
+
+#
+# Build types.hal (CryptOptions)
+#
+GEN := $(intermediates)/vendor/lineage/stache/V1_0/CryptOptions.java
+$(GEN): $(HIDL)
+$(GEN): PRIVATE_HIDL := $(HIDL)
+$(GEN): PRIVATE_DEPS := $(LOCAL_PATH)/types.hal
+$(GEN): PRIVATE_OUTPUT_DIR := $(intermediates)
+$(GEN): PRIVATE_CUSTOM_TOOL = \
+        $(PRIVATE_HIDL) -o $(PRIVATE_OUTPUT_DIR) \
+        -Ljava \
+        -randroid.hidl:system/libhidl/transport \
+        -rvendor.lineage:hardware/lineage/interfaces \
+        vendor.lineage.stache@1.0::types.CryptOptions
+
+$(GEN): $(LOCAL_PATH)/types.hal
+	$(transform-generated-source)
+LOCAL_GENERATED_SOURCES += $(GEN)
+
+#
+# Build ISecureStorage.hal
+#
+GEN := $(intermediates)/vendor/lineage/stache/V1_0/ISecureStorage.java
+$(GEN): $(HIDL)
+$(GEN): PRIVATE_HIDL := $(HIDL)
+$(GEN): PRIVATE_DEPS := $(LOCAL_PATH)/ISecureStorage.hal
+$(GEN): PRIVATE_DEPS += $(LOCAL_PATH)/types.hal
+$(GEN): $(LOCAL_PATH)/types.hal
+$(GEN): PRIVATE_OUTPUT_DIR := $(intermediates)
+$(GEN): PRIVATE_CUSTOM_TOOL = \
+        $(PRIVATE_HIDL) -o $(PRIVATE_OUTPUT_DIR) \
+        -Ljava \
+        -randroid.hidl:system/libhidl/transport \
+        -rvendor.lineage:hardware/lineage/interfaces \
+        vendor.lineage.stache@1.0::ISecureStorage
+
+$(GEN): $(LOCAL_PATH)/ISecureStorage.hal
+	$(transform-generated-source)
+LOCAL_GENERATED_SOURCES += $(GEN)
+include $(BUILD_JAVA_LIBRARY)
+
+
+################################################################################
+
+include $(CLEAR_VARS)
+LOCAL_MODULE := vendor.lineage.stache-V1.0-java-static
+LOCAL_MODULE_CLASS := JAVA_LIBRARIES
+
+intermediates := $(call local-generated-sources-dir, COMMON)
+
+HIDL := $(HOST_OUT_EXECUTABLES)/hidl-gen$(HOST_EXECUTABLE_SUFFIX)
+
+LOCAL_STATIC_JAVA_LIBRARIES := \
+    android.hidl.base-V1.0-java-static \
+
+
+#
+# Build types.hal (CryptOptions)
+#
+GEN := $(intermediates)/vendor/lineage/stache/V1_0/CryptOptions.java
+$(GEN): $(HIDL)
+$(GEN): PRIVATE_HIDL := $(HIDL)
+$(GEN): PRIVATE_DEPS := $(LOCAL_PATH)/types.hal
+$(GEN): PRIVATE_OUTPUT_DIR := $(intermediates)
+$(GEN): PRIVATE_CUSTOM_TOOL = \
+        $(PRIVATE_HIDL) -o $(PRIVATE_OUTPUT_DIR) \
+        -Ljava \
+        -randroid.hidl:system/libhidl/transport \
+        -rvendor.lineage:hardware/lineage/interfaces \
+        vendor.lineage.stache@1.0::types.CryptOptions
+
+$(GEN): $(LOCAL_PATH)/types.hal
+	$(transform-generated-source)
+LOCAL_GENERATED_SOURCES += $(GEN)
+
+#
+# Build ISecureStorage.hal
+#
+GEN := $(intermediates)/vendor/lineage/stache/V1_0/ISecureStorage.java
+$(GEN): $(HIDL)
+$(GEN): PRIVATE_HIDL := $(HIDL)
+$(GEN): PRIVATE_DEPS := $(LOCAL_PATH)/ISecureStorage.hal
+$(GEN): PRIVATE_DEPS += $(LOCAL_PATH)/types.hal
+$(GEN): $(LOCAL_PATH)/types.hal
+$(GEN): PRIVATE_OUTPUT_DIR := $(intermediates)
+$(GEN): PRIVATE_CUSTOM_TOOL = \
+        $(PRIVATE_HIDL) -o $(PRIVATE_OUTPUT_DIR) \
+        -Ljava \
+        -randroid.hidl:system/libhidl/transport \
+        -rvendor.lineage:hardware/lineage/interfaces \
+        vendor.lineage.stache@1.0::ISecureStorage
+
+$(GEN): $(LOCAL_PATH)/ISecureStorage.hal
+	$(transform-generated-source)
+LOCAL_GENERATED_SOURCES += $(GEN)
+include $(BUILD_STATIC_JAVA_LIBRARY)
+
+
+
+include $(call all-makefiles-under,$(LOCAL_PATH))
diff --git a/stache/1.0/ISecureStorage.hal b/stache/1.0/ISecureStorage.hal
new file mode 100644
index 0000000..29ac253
--- /dev/null
+++ b/stache/1.0/ISecureStorage.hal
@@ -0,0 +1,24 @@
+/*
+ * Copyright (C) 2018 The LineageOS Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package vendor.lineage.stache@1.0;
+
+interface ISecureStorage {
+    getContainerStatus(string dirPath) generates (int32_t status);
+    createContainer(string dirPath, CryptOptions options) generates (bool rc);
+    attachContainer(string dirPath, CryptOptions options) generates (bool rc);
+    detachContainer(string dirPath) generates (bool rc);
+};
diff --git a/stache/1.0/types.hal b/stache/1.0/types.hal
new file mode 100644
index 0000000..d5b73f3
--- /dev/null
+++ b/stache/1.0/types.hal
@@ -0,0 +1,26 @@
+/*
+ * Copyright (C) 2018 The LineageOS Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package vendor.lineage.stache@1.0;
+
+struct CryptOptions {
+    bool verbose;
+    string contentsCipher;
+    string filenameCipher;
+    uint32_t filenamePadding;
+    string keyDescriptor;
+    bool requiresDescriptor;
+};
diff --git a/stache/Android.bp b/stache/Android.bp
new file mode 100644
index 0000000..bbb3e4b
--- /dev/null
+++ b/stache/Android.bp
@@ -0,0 +1,4 @@
+// This is an autogenerated file, do not edit.
+subdirs = [
+    "1.0",
+]
-- 
2.17.1

